src/core/receptor.cairo

```diff
@@ -0,0 +1,299 @@
+            self.access_control.assert_has_role(receptor_roles::UPDATE_YIN_PRICE);
+            self.update_yin_price_internal();
+        }
+    }
+
+    #[abi(embed_v0)]
+    impl ITaskImpl of ITask<ContractState> {
+        fn probe_task(self: @ContractState) -> bool {
+            let seconds_since_last_update: u64 = get_block_timestamp()
+                - self.last_update_yin_price_call_timestamp.read();
+            self.update_frequency.read() <= seconds_since_last_update
+        }
+
+        fn execute_task(ref self: ContractState) {
+            assert(self.probe_task(), 'REC: Too soon to update price');
+            self.update_yin_price_internal();
+        }
+    }
+
+    //
+    // Internal Receptor functions
+    //
+
+    #[generate_trait]
+    impl ReceptorHelpers of ReceptorHelpersTrait {
+        fn set_oracle_extension_helper(ref self: ContractState, oracle_extension: ContractAddress) {
+            assert(oracle_extension.is_non_zero(), 'REC: Zero address for extension');
+
+            self.oracle_extension.write(IEkuboOracleExtensionDispatcher { contract_address: oracle_extension });
+        }
+
+        // Note that this function does not check for duplicate tokens.
+        fn set_quote_tokens_helper(ref self: ContractState, quote_tokens: Span<QuoteTokenInfo>) {
+            let mut index = LOOP_START;
+            let end_index = LOOP_START + NUM_QUOTE_TOKENS;
+            let num_quote_tokens = quote_tokens.len();
+            assert(num_quote_tokens == NUM_QUOTE_TOKENS, 'REC: Not 3 quote tokens');
+
+            let mut quote_tokens_copy = quote_tokens;
+            loop {
+                if index == end_index {
+                    break;
+                }
+                let quote_token_info: QuoteTokenInfo = *quote_tokens_copy.pop_front().unwrap();
+                self.quote_tokens.write(index, quote_token_info);
+                index += 1;
+            };
+
+            self.emit(QuoteTokensUpdated { quote_tokens });
+        }
+
+        fn set_twap_duration_helper(ref self: ContractState, twap_duration: u64) {
+            assert(twap_duration.is_non_zero(), 'REC: TWAP duration is 0');
+
+            let old_duration: u64 = self.twap_duration.read();
+            self.twap_duration.write(twap_duration);
+
+            self.emit(TwapDurationUpdated { old_duration, new_duration: twap_duration });
+        }
+
+        fn set_update_frequency_helper(ref self: ContractState, new_frequency: u64) {
+            let old_frequency: u64 = self.update_frequency.read();
+            self.update_frequency.write(new_frequency);
+            self.emit(UpdateFrequencyUpdated { old_frequency, new_frequency });
+        }
+
+        fn update_yin_price_internal(ref self: ContractState) {
+            let quotes = self.get_quotes();
+
+            let mut quotes_copy = quotes;
+            loop {
+                match quotes_copy.pop_front() {
+                    Option::Some(quote) => { if quote.is_zero() {
```

Comment:
Can this ever happen, that the oracle extension returns a 0 price? IMO it's fine to keep the code, I like the defensive style, just wondering if the situation might arise.

Reply:
It could happen if the quote token's address is wrong, so it is a sanity check on the quote tokens in a way.